### PR TITLE
US901 - P1.4 Human Jail Turn Handling

### PR DESCRIPTION
**Pull request by nicholasnastase**
Task#926: Implement jail status check and branch at turn start
Task#929: Handle turn flow continuation based on jail outcome (failed roll, successful roll, fine/card
Task#927: Build jail options UI
Task#928: Wire jail options to JailUtility

### CODE REVIEW

**Comment by dzadroga** [comment]
# Code Review Checklist

**Completed for**: US #901 and Tasks #926 to #928 
**Code Review Completed By**: Dakota Zadroga

Use this checklist to review new code changes before merging.

---

## Implementation
- [X] Does this code change accomplish what it is supposed to do?
- [X] Can the solution be simplified?
- [X] This change does **not** add unwanted compile-time or run-time dependencies.
- [X] The code is at the right abstraction level.
- [X] The code is modular enough.
- [X] The solution is adequate for maintainability, readability, performance, and security.
- [X] Similar functionality does **not** already exist in the codebase (or reuse is not appropriate).
- [X] No additional best practices or design patterns would substantially improve this code.

---

## Logic Errors and Bugs
- [X] No use cases where the code behaves incorrectly.
- [X] No inputs or external events that can break the code.
- [X] Code agrees with specifications.
- [X] No off-by-one errors.
- [X] No magic numbers (or they are well-named constants).
- [X] Proper use of global/large-scope variables (minimized and justified).

---

## Error Handling and Logging
- [X] Error handling is implemented correctly.
- [X] Error messages are user-friendly (where applicable).
- [X] Sufficient, meaningful log events exist for debugging (no sensitive data in logs).

---

## Dependencies
- [X] Documentation/configuration/README updates were made as needed.
- [X] No negative impact on other parts of the system or backward compatibility.

---

## Security and Data Privacy
- [X] No new security vulnerabilities introduced.
- [X] All user input is validated, sanitized, and escaped (XSS/SQL injection prevented).
- [X] Data from external APIs/libraries is handled safely and checked as needed.

---

## Performance
- [X] No significant performance regressions.
- [X] No obvious opportunities for significant performance gains left unaddressed.

---

## Ethics and Morality
- [X] Appropriate measures exist to prevent or report harassment/abuse (if applicable).

---

## Testing and Testability
- [X] Code is testable.
- [X] Automated tests have been added or updated.
- [X] Existing tests sufficiently cover the change (unit/integration/system).
- [X] Edge cases and boundary inputs are tested.

---

## Readability
- [X] Code is easy to understand.
- [X] Methods/functions are appropriately small and focused.
- [X] Names (functions/classes/variables) are clear and descriptive.
- [X] Code is in the correct file/folder/package.
- [X] Control flow is intuitive; data flow is clear.
- [X] Comments are clear, relevant, and up to date.
- [X] No commented-out or redundant code remains.
- [X] No overly dense one-liners; no overly long methods.

---

**Use this before submitting a pull request.**  
It helps ensure code quality, maintainability, and compliance with best practices.

*Based on awesome code review guides and MIT’s code review recommendations.*

File: Assets/Scripts/Managers/TurnFlow/TurnFlowCoordinator.cs [hunk 1]
```diff
@@ -97,7 +105,52 @@ private void OnTurnStarted(TurnStartedEvent data)
         // after dice roll, wait for movement to contine
         private void OnDiceRolled(DiceRolledEvent diceEvent)
         {
-            if (Phase != TurnPhase.AwaitingRoll) return;
+            if (Phase != TurnPhase.AwaitingRoll || diceEvent == null)
```
Comment (hdathert) [inline Assets/Scripts/Managers/TurnFlow/TurnFlowCoordinator.cs:132]: If this event is caught and checked on `OnDiceRolled` in `TurnFlowCoordinator`, then... isn't movement happening while a player is in jail?

How are we differentiating on a dice roll the context of that dice roll? Because my understanding of the current system is that `MovementStrategy` is *also* theoretically catching dice rolls.

Although, looking at it more thoroughly, `MovementStrategy` actually isn't getting called from *anywhere* right now, which I think Taylor is actively trying to solve in US899.

A stop-gap measure that might work is that we just have the movement strategy check if the player is rolling from jail, and early exit- since the movement **should not** happen while a player is in jail. `MovementStrategy` has access to `Player`, so it can check if they're in jail, right? That seems like a reasonable solution.

**Review by hdathert — CHANGES_REQUESTED** [review]
This currently does not adequately account for a player being in jail while rolling- which can/will allow player movement while they're in jail. This needs to be accounted for before approval/merge.

File: Assets/Scripts/Managers/TurnFlow/TurnFlowCoordinator.cs (hunk 1 above)
Comment (nicholasnastase) [reply Assets/Scripts/Managers/TurnFlow/TurnFlowCoordinator.cs:132]: should be resolve now, lmk if that fix works.

File: Assets/Scripts/UI/PlayerControllerUI/JailOptionsPanelController.cs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        uiActionEventChannel?.RaiseEvent(
+            new UIActionEvent(
+                UIType.JailOptions,
+                new JailActionContext(choice)));
+
+
+        Hide();
+    }
+
+    private void Show()
```
Comment (tsaldrid) [inline Assets/Scripts/UI/PlayerControllerUI/JailOptionsPanelController.cs:131]: The Show() and Hide() functions have some issues.

Because of the way OnEnable() and OnDisable() work, setting the object to inactive removes the listener, making it so it can't be used multiple times.

The way I got around this was to add a CanvasGroup to the prefab object, and then in Show() and Hide() I would use "GetComponent<CanvasGroup>.interactable = true" (or false depnending) and .alpha = 1 to show, 0 for hiding). You can look at the DicePanelController on US899 for reference.

**Review by tsaldrid — CHANGES_REQUESTED** [review]
Requesting a change to the Show() and Hide() methods in JailOptionsPanelController.cs due to reusability concerns.

File: Assets/Scripts/UI/PlayerControllerUI/JailOptionsPanelController.cs (hunk 2 above)
Comment (nicholasnastase) [reply Assets/Scripts/UI/PlayerControllerUI/JailOptionsPanelController.cs:131]: this should be good now! lmk if that's what you were thinking on that fix.

**Review by hdathert — APPROVED** [review]
This looks correct, but is hard to verify until we merge it into US899.
